src/iamserver/publicmessagehandler.cpp

```diff
@@ -0,0 +1,440 @@
+    dst.set_type(src.mType.CStr());
+    dst.set_name(src.mName.CStr());
+    dst.set_status(src.mStatus.ToString().CStr());
+    dst.set_os_type(src.mOSType.CStr());
+    dst.set_max_dmips(src.mMaxDMIPS);
+    dst.set_total_ram(src.mTotalRAM);
+
+    for (const auto& attr : src.mAttrs) {
+        ConvertToProto(attr, *dst.add_attrs());
+    }
+
+    for (const auto& partition : src.mPartitions) {
+        ConvertToProto(partition, *dst.add_partitions());
+    }
+
+    for (const auto& cpuInfo : src.mCPUs) {
+        ConvertToProto(cpuInfo, *dst.add_cpus());
+    }
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+aos::Error PublicMessageHandler::Init(NodeController& nodeController,
+    aos::iam::identhandler::IdentHandlerItf& identHandler, aos::iam::permhandler::PermHandlerItf& permHandler,
+    aos::iam::NodeInfoProviderItf& nodeInfoProvider, aos::iam::nodemanager::NodeManagerItf& nodeManager,
+    aos::iam::provisionmanager::ProvisionManagerItf& provisionManager)
+{
+    mNodeController   = &nodeController;
+    mIdentHandler     = &identHandler;
+    mPermHandler      = &permHandler;
+    mNodeInfoProvider = &nodeInfoProvider;
+    mNodeManager      = &nodeManager;
+    mProvisionManager = &provisionManager;
+
+    if (auto err = mNodeInfoProvider->GetNodeInfo(mNodeInfo); !err.IsNone()) {
+        LOG_ERR() << "Failed to get node info: " << AOS_ERROR_WRAP(err);
+
+        return AOS_ERROR_WRAP(err);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+void PublicMessageHandler::RegisterServices(grpc::ServerBuilder& builder)
+{
+    builder.RegisterService(static_cast<iamanager::IAMVersionService::Service*>(this));
+    builder.RegisterService(static_cast<iamproto::IAMPublicService::Service*>(this));
+
+    if (GetPermHandler() != nullptr) {
+        builder.RegisterService(static_cast<iamproto::IAMPublicPermissionsService::Service*>(this));
+    }
+
+    if (IsMainNode()) {
+        if (GetIdentHandler() != nullptr) {
+            builder.RegisterService(static_cast<iamproto::IAMPublicIdentityService::Service*>(this));
+        }
+
+        builder.RegisterService(static_cast<iamproto::IAMPublicNodesService::Service*>(this));
+    }
+}
+
+void PublicMessageHandler::OnNodeInfoChange(const aos::NodeInfo& info)
+{
+    LOG_DBG() << "On node info changed: ID = " << info.mID;
+
+    iamproto::NodeInfo nodeInfo;
+    ConvertToProto(info, nodeInfo);
+
+    mNodeChangedController.WriteToSteams(nodeInfo);
+}
+
+void PublicMessageHandler::OnNodeRemoved(const aos::String& id)
+{
+    LOG_DBG() << "On node removed: ID = " << id;
+}
+
+aos::Error PublicMessageHandler::SubjectsChanged(const aos::Array<aos::StaticString<aos::cSubjectIDLen>>& messages)
+{
+    LOG_DBG() << "Subjects changed: count = " << messages.Size();
```

Comment:
We are not so interested in count actually. Either log all subjects or remove count.